ckan/config/middleware/__init__.py

```diff
@@ -127,8 +127,21 @@ def handle_i18n(self, environ):
             * CKAN_CURRENT_URL is set to the current application url
         '''
 
+        # Ensure URL encoding
+        for var in ('PATH_INFO', 'REQUEST_URI', 'QUERY_STRING'):
+            val = environ.get(var)
+            if val:
+                try:
+                    val.decode('utf-8')
+                except UnicodeDecodeError:
+                    val = val.decode('cp1252').encode('utf-8')
+
+                # Preserve reserved characters and avoid double-quoting
+                environ[var] = urllib.quote(val, '%:;/?&=@')
```

Comment:
what's the quoting here for?

Reply:
generally values inside the application to be Unicode and unquoted for any particular use. That makes handling them less surprising. When values are sent outside the application they should be encoded/quoted based on where they're being sent.